// @tma/sdk/src/index.ts
/**
 * Bridge.
 */
export { createPostEvent } from './bridge/methods/createPostEvent.js';
export { invokeCustomMethod } from './bridge/invokeCustomMethod.js';
export { on } from './bridge/events/on.js';
export { off } from './bridge/events/off.js';
export { once } from './bridge/events/once.js';
export { parseMessage } from './bridge/parseMessage.js';
export { postEvent, type PostEvent } from './bridge/methods/postEvent.js';
export { request, type RequestOptions, type RequestOptionsAdvanced } from './bridge/request.js';
export { subscribe } from './bridge/events/subscribe.js';
export { unsubscribe } from './bridge/events/unsubscribe.js';
export { MethodUnsupportedError } from './bridge/errors/MethodUnsupportedError.js';
export { ParameterUnsupportedError } from './bridge/errors/ParameterUnsupportedError.js';
export type {
  HeaderColorKey,
  MiniAppsEmptyMethodName,
  MiniAppsMethodAcceptParams,
  MiniAppsMethodName,
  MiniAppsMethodParams,
  MiniAppsMethods,
  MiniAppsNonEmptyMethodName,
  SwitchInlineQueryChatType,
} from './bridge/methods/methods.ts';
export type { InvoiceStatus } from './bridge/events/parsers/invoiceClosed.ts';
export type {
  ImpactHapticFeedbackStyle,
  NotificationHapticFeedbackType,
} from './bridge/methods/haptic.ts';
export type {
  MiniAppsEventEmitter, MiniAppsEventHasParams,
  MiniAppsEventListener,
  MiniAppsEventName,
  MiniAppsEventParams, MiniAppsEvents, MiniAppsGlobalEventListener,
} from './bridge/events/events.ts';
export type { PhoneRequestedStatus } from './bridge/events/parsers/phoneRequested.ts';
export type { WriteAccessRequestedStatus } from './bridge/events/parsers/writeAccessRequested.ts';

/**
 * Classnames.
 */
export { classNames } from './classnames/classNames.js';
export { mergeClassNames } from './classnames/mergeClassNames.js';

/**
 * Colors.
 */
export { isRGB } from './colors/isRGB.js';
export { isRGBShort } from './colors/isRGBShort.js';
export { isColorDark } from './colors/isColorDark.js';
export { toRGB } from './colors/toRGB.js';
export type { RGB, RGBShort } from './colors/types.ts';

/**
 * Components.
 */

// Closing behavior
export { ClosingBehavior } from './components/closing-behavior/ClosingBehavior.js';
export type {
  ClosingBehaviorEventListener,
  ClosingBehaviorEventName,
  ClosingBehaviorEvents,
} from './components/closing-behavior/types.ts';

// Cloud storage.
export { CloudStorage } from './components/cloud-storage/CloudStorage.js';

// HapticFeedback.
export { HapticFeedback } from './components/haptic-feedback/HapticFeedback.js';

// Init data.
export { chatParser } from './components/init-data/chatParser.js';
export { InitData } from './components/init-data/InitData.js';
export { initDataParser } from './components/init-data/initDataParser.js';
export { parseInitData } from './components/init-data/parseInitData.js';
export { userParser } from './components/init-data/userParser.js';
export type { Chat, User, ChatType, InitDataParsed } from './components/init-data/types.ts';

// Invoice.
export { Invoice } from './components/invoice/Invoice.js';
export type {
  InvoiceState,
  InvoiceEventListener,
  InvoiceEventName,
  InvoiceEvents,
} from './components/invoice/types.ts';

// Main button.
export { MainButton } from './components/main-button/MainButton.js';
export type {
  MainButtonParams,
  MainButtonEvents,
  MainButtonEventListener,
  MainButtonEventName,
  MainButtonProps,
} from './components/main-button/types.ts';

// Mini app.
export { MiniApp } from './components/mini-app/MiniApp.js';
export type {
  MiniAppHeaderColor,
  MiniAppProps,
  MiniAppEvents,
  MiniAppEventListener,
  MiniAppEventName,
} from './components/mini-app/types.ts';

// Popup.
export { Popup } from './components/popup/Popup.js';
export type {
  PopupEventName,
  PopupEventListener,
  OpenPopupOptions,
  OpenPopupOptionsButton,
  PopupEvents,
} from './components/popup/types.ts';

// QR scanner.
export { QRScanner } from './components/qr-scanner/QRScanner.js';
export type {
  QRScannerEvents,
  QRScannerEventListener,
  QRScannerEventName,
} from './components/qr-scanner/types.ts';

// Settings button.
export { SettingsButton } from './components/settings-button/SettingsButton.js';
export type {
  SettingsButtonEventListener,
  SettingsButtonEventName,
  SettingsButtonEvents,
} from './components/settings-button/types.ts';

// Theme params.
export { parseThemeParams } from './components/theme-params/parseThemeParams.js';
export { requestThemeParams } from './components/theme-params/requestThemeParams.js';
export { serializeThemeParams } from './components/theme-params/serializeThemeParams.js';
export { themeParamsParser } from './components/theme-params/themeParamsParser.js';
export { ThemeParams } from './components/theme-params/ThemeParams.js';
export type {
  ThemeParamsEventListener,
  ThemeParamsEventName,
  ThemeParamsKey,
  ThemeParamsEvents,
  ThemeParamsParsed,
} from './components/theme-params/types.ts';

// Utils.
export { Utils } from './components/utils/Utils.js';

// Viewport.
export { isStableViewportPlatform } from './components/viewport/isStableViewportPlatform.js';
export {
  requestViewport,
  type RequestViewportResult,
} from './components/viewport/requestViewport.js';
export { Viewport } from './components/viewport/Viewport.js';
export type {
  ViewportProps,
  ViewportEventListener,
  ViewportEventName,
  ViewportEvents,
} from './components/viewport/types.ts';

/**
 * CSS.
 */
export { setCSSVar } from './css/setCSSVar.js';
export { bindMiniAppCSSVars } from './css/bindMiniAppCSSVars.js';
export { bindThemeCSSVars } from './css/bindThemeCSSVars.js';
export { bindViewportCSSVars } from './css/bindViewportCSSVars.js';

/**
 * Init.
 */
export { init } from './init/init.js';
export type { InitOptions, InitResult } from './init/types.ts';

/**
 * Launch params.
 */
export { launchParamsParser } from './launch-params/launchParamsParser.js';
export { parseLaunchParams } from './launch-params/parseLaunchParams.js';
export { retrieveLaunchParams } from './launch-params/retrieveLaunchParams.js';
export { retrieveLaunchData } from './launch-params/retrieveLaunchData.js';
export { serializeLaunchParams } from './launch-params/serializeLaunchParams.js';
export type { LaunchParams } from './launch-params/types.ts';
export type { LaunchData } from './launch-params/types.ts';

/**
 * Misc.
 */
export { isTMA } from './misc/isTMA.js';
export { isRecord } from './misc/isRecord.js';
export { isIframe } from './misc/isIframe.js';
export { isPageReload } from './misc/isPageReload.js';

/**
 * Navigation.
 */
export { getHash } from './navigation/getHash.js';
export { HashNavigator } from './navigation/HashNavigator/HashNavigator.js';
export { Navigator } from './navigation/Navigator/Navigator.js';
export type {
  NavigationEntry,
  NavigatorConEntry,
  NavigatorOptions,
} from './navigation/Navigator/types.ts';
export type {
  HashNavigatorOptions,
  HashNavigatorEventListener,
  HashNavigatorEventsMap,
  HashNavigatorEventName,
} from './navigation/HashNavigator/types.ts';

/**
 * Parsing.
 */
export { boolean } from './parsing/parsers/boolean.js';
export { searchParams } from './parsing/parsers/searchParams.js';
export { string } from './parsing/parsers/string.js';
export { rgb } from './parsing/parsers/rgb.js';
export { array } from './parsing/parsers/array.js';
export { date } from './parsing/parsers/date.js';
export { json } from './parsing/parsers/json.js';
export { number } from './parsing/parsers/number.js';
export { ParseError } from './parsing/ParseError.js';
export { ParseSchemaFieldError } from './parsing/ParseSchemaFieldError.js';

/**
 * Supports.
 */
export { supports } from './supports/supports.js';

/**
 * Timeout.
 */
export { withTimeout } from './timeout/withTimeout.js';
export { TimeoutError } from './timeout/TimeoutError.js';
export { isTimeoutError } from './timeout/isTimeoutError.js';

/**
 * Types.
 */
export type { RequestId, CreateRequestIdFunc } from './types/request-id.ts';

/**
 * Version.
 */
export { compareVersions } from './version/compareVersions.js';
export type { Version } from './version/types.ts';

/**
 * Globals.
 */
export { setTargetOrigin, setDebug } from './globals.js';
